qmp/libvirtgo_linux.go

```diff
@@ -0,0 +1,327 @@
+
+	if mon.virConn == nil {
+		return nil,
+			errors.New("Events() need a established connection to proceed.")
+	}
+
+	callbackID, err := domainEventRegister(mon)
+	if err != nil {
+		return nil, err
+	}
+
+	go pollEventsLoop(mon.doneChan)
+
+	go domainEventDeregister(mon, callbackID)
+
+	return mon.eventsChan, nil
+}
+
+var lookupDomainByName = func(
+	virConn *libvirt.VirConnection,
+	domainName string) (libvirt.VirDomain, error) {
+	return virConn.LookupDomainByName(domainName)
+}
+
+// eventRegisterDefaultImpl registers a default event implementation.
+func eventRegisterDefaultImpl() {
+	eventRegisterID := eventRegisterDefaultImplInternal()
+	if eventRegisterID == -1 {
+		log.Printf(
+			"EventRegisterDefaultImpl returned an unexpected value %d\n",
+			eventRegisterID)
+		//TODO: panic or what???
+	}
+}
+
+var eventRegisterDefaultImplInternal = func() int {
+	return libvirt.EventRegisterDefaultImpl()
+}
+
+// domainEventRegister register with libvirt to receive events for the
+// specified domain.
+func domainEventRegister(mon *libvirtGoMonitorLinux) (int, error) {
+	callback :=
+		libvirt.DomainEventCallback(newEventCallback(mon))
+
+	//TODO: add ability to register for more event types
+	callbackID := domainEventRegisterInternal(
+		mon,
+		&callback,
+		func() {
+			// empty on purpose until it's decided
+			// what to do here.
+		},
+	)
+	if callbackID == -1 {
+		return -1, errors.New("Domain event registration failed!")
+	}
+
+	return callbackID, nil
+}
+
+var domainEventRegisterInternal = func(
+	mon *libvirtGoMonitorLinux,
+	callback *libvirt.DomainEventCallback,
+	fn func()) int {
+	return mon.virConn.DomainEventRegister(
+		*mon.domain,
+		libvirt.VIR_DOMAIN_EVENT_ID_LIFECYCLE,
+		callback,
+		fn,
+	)
+}
+
+// domainEventDeregister stops the registration with libvirt from receiving
+// events for the specified domain.
+func domainEventDeregister(
+	mon *libvirtGoMonitorLinux,
+	callbackID int) {
+	<-mon.doneChan
+	ret := domainEventDeregisterInternal(mon.virConn, callbackID)
+	if ret != 0 {
+		log.Printf("DomainEventDeregister returned an unexpected value: %d\n",
+			ret)
+	}
+}
+
+var domainEventDeregisterInternal = func(virConn *libvirt.VirConnection,
+	callbackID int) int {
+	return virConn.DomainEventDeregister(callbackID)
+}
+
+// newEventCallback convenient function to provide access
+// to the eventChan to the returned closure/callback.
+func newEventCallback(mon *libvirtGoMonitorLinux) libvirt.DomainEventCallback {
+	return func(c *libvirt.VirConnection,
+		d *libvirt.VirDomain,
+		eventDetails interface{}, f func()) int {
+
+		// if monitor is not connected,
+		// we should not continue processing messages
+		// as the mon.eventsChan will be closed.
+		if mon.virConn == nil {
+			return 0
+		}
+
+		// We only support lifecycleEvents for now
+		if lifecycleEvent, ok :=
+			eventDetails.(libvirt.DomainLifecycleEvent); ok {
+			mon.eventsChan <- constructEvent(lifecycleEvent)
+			f()
+		}
+
+		// according to the libvirt-do
```

Comment:
it's a typo. It should read "according to the libvirt-go documentation the return code from DomainEventCallback is ignored."